langchain4j/src/main/java/dev/langchain4j/memory/chat/SummarizingChatMemory.java

```diff
@@ -0,0 +1,340 @@
+package dev.langchain4j.memory.chat;
+
+import static dev.langchain4j.internal.ValidationUtils.ensureGreaterThanZero;
+import static dev.langchain4j.internal.ValidationUtils.ensureNotNull;
+
+import dev.langchain4j.data.message.AiMessage;
+import dev.langchain4j.data.message.ChatMessage;
+import dev.langchain4j.data.message.SystemMessage;
+import dev.langchain4j.data.message.ToolExecutionResultMessage;
+import dev.langchain4j.memory.ChatMemory;
+import dev.langchain4j.model.chat.ChatModel;
+import dev.langchain4j.model.chat.response.ChatResponse;
+import dev.langchain4j.service.memory.ChatMemoryService;
+import dev.langchain4j.store.memory.chat.ChatMemoryStore;
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Function;
+
+/**
+ * A chat memory implementation that automatically summarizes conversation history
+ * when the number of messages exceeds a configurable threshold.
```

Comment:
```suggestion
 * when the number of messages exceeds a configurable `maxMessages`.
```
Otherwise it can be confused with `summarizeThreshold`

Reply:
But in general I think it would be more useful to have `maxTokens` setting instead of `maxMessages`, as it is much more precise and controllable

Reply:
Indeed, maxTokens would be better. It can prevent the summary from exceeding the token limit of the large model due to an excessive number of messages. I will switch to using tokens to summarize the conversation.

Reply:
However, if we switch to using maxTokens, it becomes difficult to control whether the generated summary exceeds maxTokens. Therefore, I think we can consider providing an interface function, allowing users to summarize according to their own strategies if they have any.